File: python/48_Rotate_Image.py
class Solution:
    # @param {integer[][]} matrix
    # @return {void} Do not return anything, modify matrix in-place instead.
    def rotate(self, matrix):
        if not matrix:
            return
        degree=len(matrix)
        for layer in xrange(degree/2):
            first=layer
            last=degree-1-layer
            for i in xrange(first,last):
                offset=i-first
                top=matrix[first][i]
                matrix[layer][i]=matrix[last-offset][first]
                matrix[last-offset][first]=matrix[last][last-offset]
                matrix[last][last-offset]=matrix[i][last]
                matrix[i][last]=top

File: Java/48_Rotate_Image.java
public class Solution {
    // 旋转图形矩阵由于要访问所有的元素，所以问题的时间复杂度为O(n^2)，没法降低
    // 本方法就是最直接的模拟法，从最外围到内一层层旋转
    // 本题唯一要注意的地方是下标的变换关系
    public void rotate(int[][] matrix) {
        if(matrix==null||matrix.length<=1||matrix[0].length<=1)return;
        int top=0, left=0, right=matrix.length-1, bottom=matrix.length-1;
        int n=matrix.length;
        int i, keep;
        while(left<right){
            for(i=left;i<right;i++){
                keep=matrix[top][i];
                matrix[top][i]=matrix[n-i-1][left];
                matrix[n-i-1][left]=matrix[bottom][n-i-1];
                matrix[bottom][n-i-1]=matrix[i][right];
                matrix[i][right]=keep;
            }
            top++;
            bottom--;
            left++;
            right--;
        }
    }
}